shell/platform/android/vsync_waiter_android.cc

```diff
@@ -57,6 +57,8 @@ void VsyncWaiterAndroid::OnNativeVsync(JNIEnv* env,
                                        jlong frameTimeNanos,
                                        jlong frameTargetTimeNanos,
                                        jlong java_baton) {
+  TRACE_EVENT0("flutter", "VSYNC");
+
```

Comment:
The time of the actual VSYNC was given in the arguments above as "frameTimeNanos". This TRACE call is recording the current wall time as the VSYNC event which means the gray bars are not accurately showing the platform/hardware vertical sync windows.

Reply:
FML tracing library doesn't seem to expose a facility to trace with an explicit timestamp:
https://github.com/flutter/engine/blob/master/fml/trace_event.h

What would be the right thing to do?

Reply:
If you trace the macros down to the eventual function called I believe that it allows a timestamp to be included, but it's been a while since I was playing with that mechanism. There are async versions of the macros that do take timestamps and one of these is in use in the call in the vsync_waiter to record the VsyncSchedulingOverhead event, but I'm not sure if the tracing observatories allow the VSYNC events to be async versions (which use different event codes). Some experimentation would be required with Async/non-Async and possibly some new macros or code in fml::tracing.

Reply:
1. Tried making VSYNC an async event with explicit timestamps:
![async-VSYNC-async-trace-from-devtools](https://user-images.githubusercontent.com/720390/75342484-d29c0380-5896-11ea-955b-8c213512e8d5.png)
https://drive.google.com/a/google.com/file/d/1KogxMavJW9tYQhzBwvu9Xqhsm90iF0Lb/view?usp=sharing

This is not appropriate: async traces get displayed outside of their threads and do not get stripe highlighting.

2. Sending sync events with explicit timestamps would violate the [trace format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit#):

> The timestamps for the duration events must be in increasing order for a given thread.

and in practice will generate [a warning from the loader](https://github.com/catapult-project/catapult/blob/17f4e8b9da14d3b07f7adf39e8957804bfa46dd0/tracing/tracing/extras/importer/linux_perf/android_parser.html#L120).

Reply:
I think I ran into the same problems when I tried to do something similar a few months ago. We would have to connect with the team that writes the tracing code and work out something to really fix this. If they could just allow the VSYNC events to be delivered as async events then we could achieve what we're trying to do.

In the meantime, if this fix is blocking somebody maybe we should start with what you have here and work on getting the VSYNC bars aligned correctly separately as a future PR. As long as the VsyncSchedulingOverhead async event is still shown on the output there is at least a way to track the missing time.

One thing I might suggest is moving the VsyncOverhead event into the platform waiters if it isn't needed for the other cases. Then if/when we get the observatory to allow asynchronous VSYNC we can eliminate the 2 events in favor of the one.

Reply:
Changes to trace format are a bit complicated since there are three implementations of trace viewers:
1) Dart Devtools,
2) Chrome Tracing,
3) http://perfetto.dev.

It's especially tricky, since Chrome Trace viewer team moved on to work on perfetto.dev and the original trace viewer is not maintained.

Moving VsyncSchedulingOverhead is tempting, but would need to be duplicated in all of the platform embedders, including platform ones. At the same time, as is it is reflecting correct data on all platforms despite being in the common code and doesn't mess up trace rendering.

So for now I'm inclined to keep this change as it is.